BUG preserve_key_order still method not found

**Describe the bug**
Half a year later, I still cannot `cargo install qsv -...`.

**To Reproduce**
Steps to reproduce the behavior:
1. `cargo install qsv --locked --features feature_capable`

**Expected behavior**
It should produce a `qsv` binary.

**Screenshots/Backtrace/Sample Data**
```text
   Compiling qsv v16.1.0
error[E0599]: no method named `preserve_key_order` found for struct `Json2Csv` in the current scope
   --> /home/home/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/qsv-16.1.0/src/cmd/json.rs:330:14
    |
329 | /         Json2Csv::new(flattener)
330 | |             .preserve_key_order(true)
    | |             -^^^^^^^^^^^^^^^^^^ method not found in `Json2Csv`
    | |_____________|
    |

For more information about this error, try `rustc --explain E0599`.
error: could not compile `qsv` (bin "qsv") due to 1 previous error
error: failed to compile `qsv v16.1.0`, intermediate artifacts can be found at `/tmp/cargo-installohKHyA`.
To reuse those artifacts with a future compilation, set the environment variable `CARGO_TARGET_DIR` to that path.
```

**Desktop (please complete the following information):**
 - OS: Ubuntu 25.10
 - qsv v16.1.0

**Additional context**

You [said](https://github.com/dathere/qsv/issues/2819#issuecomment-3102267604):

> If its not merged, I may ultimately make a renamed fork and publish it on crates.io as I've done with other unmaintained crates like dynfmt, docopt, vader_sentiment_analysis, etc.

It's still not merged after so long, and that dependency is the one thing holding `cargo install` back. @vtselfa hasn't been active at all last year, so I don't think it will be merged. Please fork.

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -228,7 +228,6 @@ itoa = "1"
 jaq-core = "3"
 jaq-json = { version = "2", features = ["serde"] }
 jaq-std = "3"
-json-objects-to-csv = "0.1.3"
 jsonschema = { version = "0.46", features = [
     "resolve-file",
     "resolve-http",
@@ -437,10 +436,7 @@ csvlens = { git = "https://github.com/jqnatividad/csvlens", branch = "bump_tui-i
 # use our patched fork of csvs_convert to bump dependencies until our PR is merged
 csvs_convert = { git = "https://github.com/jqnatividad/csvs_convert", branch = "replace-streaming-stats_with_qsv-stats", optional = true }
 
-# use our patched fork of json-objects-to-csv to bump deps and to preserve order until our PR is merged
-json-objects-to-csv = { git = "https://github.com/jqnatividad/json-objects-to-csv", branch = "preserve_order_issue_10" }
-
-# use latest upstream now that our PR has been merged while awaiting a new release 
+# use latest upstream now that our PR has been merged while awaiting a new release
 self_update = { git = "https://github.com/jaemk/self_update", rev = "cad8bb9" }
 
 # use our patched fork of sled to get rid of unmaintained instant

--- a/src/cmd/json.rs
+++ b/src/cmd/json.rs
@@ -148,7 +148,7 @@ Common options:
 "#;
 
 use std::{
-    collections::HashSet,
+    collections::HashMap,
     env,
     io::{Read, Write},
 };
@@ -410,27 +410,27 @@ pub fn run(argv: &[&str]) -> CliResult<()> {
     Ok(final_csv_wtr.flush()?)
 }
 
-/// Internal key separator used while flattening. It is the Unicode "group separator"
-/// control character (U+241D), chosen because it is extremely unlikely to appear in real input.
-/// Flattening with this separator (instead of the user-facing `.`) lets us detect collisions
-/// between genuinely-nested keys and keys that already literally contain a `.` (e.g.
-/// `{"a": {"b": 1}, "a.b": 2}`) before the separators are normalized back to `.`.
-const FLATTEN_INTERNAL_SEP: char = '\u{241D}';
 /// User-facing key separator for flattened nested keys (e.g. `parent.child`, `arr.0`).
 const FLATTEN_SEP: &str = ".";
 
-/// Flatten a JSON `Value` into `out`, mirroring the default behavior of the (now-removed,
-/// unmaintained) `json-objects-to-csv` / `flatten-json-object` crates:
-/// - nested object/array keys are joined with `FLATTEN_INTERNAL_SEP`,
-/// - array elements use their index as the key segment (`arr.0`, `arr.1`, ...),
-/// - empty objects and arrays are dropped,
+/// Flatten a JSON `Value` into `out` as `(path_segments, leaf_value)` pairs, in document order,
+/// mirroring the default behavior of the (now-removed, unmaintained) `json-objects-to-csv` /
+/// `flatten-json-object` crates:
+/// - each nesting level contributes one path segment (an object key, or an array index),
+/// - empty objects and arrays are dropped (they contribute no leaf),
 /// - empty-string keys and JSON `null` values are preserved,
 /// - the top-level value must be an object.
+///
+/// Keeping the path as a `Vec<String>` of raw segments (rather than a pre-joined string with a
+/// magic separator) lets a later join with `FLATTEN_SEP` produce the display header while still
+/// distinguishing structurally-different paths that happen to join to the same header (e.g. a
+/// genuinely nested `a`→`b` vs. a literal key `"a.b"`), with no risk of corrupting keys that
+/// literally contain the separator.
 fn flatten_json_value(
     current: &Value,
-    parent_key: String,
+    path: &mut Vec<String>,
     depth: u32,
-    out: &mut serde_json::Map<String, Value>,
+    out: &mut Vec<(Vec<String>, Value)>,
 ) -> CliResult<()> {
     if depth == 0 && !current.is_object() {
         return fail_clierror!("Flattening error: top-level JSON value must be an object");
@@ -439,31 +439,21 @@ fn flatten_json_value(
     if let Some(obj) = current.as_object() {
         // empty objects are not preserved (nothing to iterate)
         for (k, v) in obj {
-            let child_key = if depth > 0 {
-                format!("{parent_key}{FLATTEN_INTERNAL_SEP}{k}")
-            } else {
-                k.clone()
-            };
-            flatten_json_value(v, child_key, depth + 1, out)?;
+            path.push(k.clone());
+            flatten_json_value(v, path, depth + 1, out)?;
+            path.pop();
         }
     } else if let Some(arr) = current.as_array() {
         // empty arrays are not preserved (nothing to iterate)
         for (i, v) in arr.iter().enumerate() {
-            let child_key = format!("{parent_key}{FLATTEN_INTERNAL_SEP}{i}");
-            flatten_json_value(v, child_key, depth + 1, out)?;
+            path.push(i.to_string());
+            flatten_json_value(v, path, depth + 1, out)?;
+            path.pop();
         }
     } else {
-        if out.contains_key(&parent_key) {
-            return fail_clierror!("Flattening error: key '{parent_key}' would be overwritten");
-        }
-        out.insert(parent_key, current.clone());
+        out.push((path.clone(), current.clone()));
     }
     Ok(())
-}
-
-/// Normalize an internally-flattened key back to the user-facing `.` separator.
-fn transform_flattened_key(key: &str) -> String {
-    key.replace(FLATTEN_INTERNAL_SEP, FLATTEN_SEP)
 }
 
 /// Convert a slice of JSON `Value`s (each expected to be an object) into CSV, writing to
@@ -482,21 +472,31 @@ fn json_objects_to_csv<W: Write>(
         return Ok(());
     }
 
-    // First pass: collect the header union in first-seen order, and detect collisions that
-    // would occur when the internal separator is normalized back to `.` (e.g. a genuinely
-    // nested `a.b` key colliding with a literal "a.b" key).
+    // First pass: collect the header union in first-seen order, mapping each display header to the
+    // structural path that produced it. If a second, structurally-different path normalizes to an
+    // already-seen header, that's a collision (e.g. nested `a`→`b` vs. a literal `"a.b"` key).
     let mut headers: Vec<String> = Vec::new();
-    let mut seen_headers: HashSet<String> = HashSet::new();
-    let mut orig_keys: HashSet<String> = HashSet::new();
-    let mut flat = serde_json::Map::new();
+    let mut header_paths: HashMap<String, Vec<String>> = HashMap::new();
+    let mut leaves: Vec<(Vec<String>, Value)> = Vec::new();
+    let mut path: Vec<String> = Vec::new();
     for obj in objects {
-        flat.clear();
-        flatten_json_value(obj, String::new(), 0, &mut flat)?;
-        for orig_key in flat.keys() {
-            orig_keys.insert(orig_key.clone());
-            let header = transform_flattened_key(orig_key);
-            if seen_headers.insert(header.clone()) {
-                headers.push(header);
+        leaves.clear();
+        path.clear();
+        flatten_json_value(obj, &mut path, 0, &mut leaves)?;
+        for (segments, _) in &leaves {
+            let header = segments.join(FLATTEN_SEP);
+            match header_paths.get(&header) {
+                Some(existing) if existing != segments => {
+                    return fail_clierror!(
+                        "Flattening Key Collision error: two different JSON keys collide after \
+                         flattening to '{header}'"
+                    );
+                },
+                Some(_) => {},
+                None => {
+                    header_paths.insert(header.clone(), segments.clone());
+                    headers.push(header);
+                },
             }
         }
     }
@@ -505,30 +505,25 @@ fn json_objects_to_csv<W: Write>(
     if headers.is_empty() {
         return Ok(());
     }
-    // if two distinct flattened keys normalize to the same header, it's a collision
-    if headers.len() != orig_keys.len() {
-        return fail_clierror!(
-            "Flattening Key Collision error: two different JSON keys collide after flattening"
-        );
-    }
 
     csv_writer.write_record(&headers)?;
 
     // Second pass: write one CSV row per object.
     let mut record: Vec<String> = Vec::with_capacity(headers.len());
-    let mut transformed: serde_json::Map<String, Value> = serde_json::Map::new();
+    let mut row: HashMap<String, Value> = HashMap::new();
     for obj in objects {
-        flat.clear();
-        flatten_json_value(obj, String::new(), 0, &mut flat)?;
+        leaves.clear();
+        path.clear();
+        flatten_json_value(obj, &mut path, 0, &mut leaves)?;
 
-        transformed.clear();
-        for (orig_key, val) in &flat {
-            transformed.insert(transform_flattened_key(orig_key), val.clone());
+        row.clear();
+        for (segments, val) in leaves.drain(..) {
+            row.insert(segments.join(FLATTEN_SEP), val);
         }
 
         record.clear();
         for header in &headers {
-            match transformed.get(header) {
+            match row.get(header) {
                 Some(Value::String(s)) => record.push(s.clone()),
                 Some(v @ (Value::Bool(_) | Value::Number(_))) => record.push(v.to_string()),
                 // Null, plus any empty Array/Object that slipped through, become empty fields.

--- a/src/cmd/json.rs
+++ b/src/cmd/json.rs
@@ -147,13 +147,17 @@ Common options:
     -o, --output <file>    Write output to <file> instead of stdout.
 "#;
 
-use std::{env, io::Read};
+use std::{
+    collections::HashSet,
+    env,
+    io::{Read, Write},
+};
 
 use jaq_core::{Compiler, Ctx, Vars, data, load, unwrap_valr};
 use jaq_json::{Num, Val};
-use json_objects_to_csv::{Json2Csv, flatten_json_object::Flattener};
 use log::warn;
 use serde::Deserialize;
+use serde_json::Value;
 
 use crate::{CliError, CliResult, config, select::SelectColumns, util};
 
@@ -163,27 +167,6 @@ struct Args {
     flag_jaq:    Option<String>,
     flag_select: Option<SelectColumns>,
     flag_output: Option<String>,
-}
-
-impl From<json_objects_to_csv::Error> for CliError {
-    fn from(err: json_objects_to_csv::Error) -> Self {
-        match err {
-            json_objects_to_csv::Error::Flattening(err) => {
-                CliError::Other(format!("Flattening error: {err}"))
-            },
-            json_objects_to_csv::Error::FlattenedKeysCollision => {
-                CliError::Other(format!("Flattening Key Collision error: {err}"))
-            },
-            json_objects_to_csv::Error::WrittingCSV(err) => {
-                CliError::Other(format!("Writing CSV error: {err}"))
-            },
-            json_objects_to_csv::Error::ParsingJson(err) => {
-                CliError::Other(format!("Parsing JSON error: {err}"))
-            },
-            json_objects_to_csv::Error::InputOutput(err) => CliError::Io(err),
-            json_objects_to_csv::Error::IntoFile(err) => CliError::Io(err.into()),
-        }
-    }
 }
 
 pub fn run(argv: &[&str]) -> CliResult<()> {
@@ -370,17 +353,15 @@ pub fn run(argv: &[&str]) -> CliResult<()> {
             None => std::slice::from_ref(&value),
         };
 
-        let flattener = Flattener::new();
         let intermediate_csv_writer = csv::WriterBuilder::new()
             .has_headers(true)
             .from_path(intermediate_csv.clone())?;
-        Json2Csv::new(flattener)
-            .preserve_key_order(true)
-            .convert_from_array(values, intermediate_csv_writer)?;
+        json_objects_to_csv(values, intermediate_csv_writer)?;
     }
 
     // STEP 2: select the columns to use in the final output
-    // Read the intermediate CSV to get the actual headers (which are sorted alphabetically)
+    // Read the intermediate CSV to get the actual headers (which, since we preserve key order,
+    // appear in the order the keys were first encountered in the JSON data)
     let sel_rconfig = config::Config::new(Some(intermediate_csv).as_ref()).no_headers(false);
     let mut intermediate_csv_rdr = sel_rconfig.reader()?;
     let byteheaders = intermediate_csv_rdr.byte_headers()?;
@@ -427,6 +408,138 @@ pub fn run(argv: &[&str]) -> CliResult<()> {
     }
 
     Ok(final_csv_wtr.flush()?)
+}
+
+/// Internal key separator used while flattening. It is the Unicode "group separator"
+/// control character (U+241D), chosen because it is extremely unlikely to appear in real input.
+/// Flattening with this separator (instead of the user-facing `.`) lets us detect collisions
+/// between genuinely-nested keys and keys that already literally contain a `.` (e.g.
+/// `{"a": {"b": 1}, "a.b": 2}`) before the separators are normalized back to `.`.
+const FLATTEN_INTERNAL_SEP: char = '\u{241D}';
+/// User-facing key separator for flattened nested keys (e.g. `parent.child`, `arr.0`).
+const FLATTEN_SEP: &str = ".";
+
+/// Flatten a JSON `Value` into `out`, mirroring the default behavior of the (now-removed,
+/// unmaintained) `json-objects-to-csv` / `flatten-json-object` crates:
+/// - nested object/array keys are joined with `FLATTEN_INTERNAL_SEP`,
+/// - array elements use their index as the key segment (`arr.0`, `arr.1`, ...),
+/// - empty objects and arrays are dropped,
+/// - empty-string keys and JSON `null` values are preserved,
+/// - the top-level value must be an object.
+fn flatten_json_value(
+    current: &Value,
+    parent_key: String,
+    depth: u32,
+    out: &mut serde_json::Map<String, Value>,
+) -> CliResult<()> {
+    if depth == 0 && !current.is_object() {
+        return fail_clierror!("Flattening error: top-level JSON value must be an object");
+    }
+
+    if let Some(obj) = current.as_object() {
+        // empty objects are not preserved (nothing to iterate)
+        for (k, v) in obj {
+            let child_key = if depth > 0 {
+                format!("{parent_key}{FLATTEN_INTERNAL_SEP}{k}")
+            } else {
+                k.clone()
+            };
+            flatten_json_value(v, child_key, depth + 1, out)?;
+        }
+    } else if let Some(arr) = current.as_array() {
+        // empty arrays are not preserved (nothing to iterate)
+        for (i, v) in arr.iter().enumerate() {
+            let child_key = format!("{parent_key}{FLATTEN_INTERNAL_SEP}{i}");
+            flatten_json_value(v, child_key, depth + 1, out)?;
+        }
+    } else {
+        if out.contains_key(&parent_key) {
+            return fail_clierror!("Flattening error: key '{parent_key}' would be overwritten");
+        }
+        out.insert(parent_key, current.clone());
+    }
+    Ok(())
+}
+
+/// Normalize an internally-flattened key back to the user-facing `.` separator.
+fn transform_flattened_key(key: &str) -> String {
+    key.replace(FLATTEN_INTERNAL_SEP, FLATTEN_SEP)
+}
+
+/// Convert a slice of JSON `Value`s (each expected to be an object) into CSV, writing to
+/// `csv_writer`. Each object becomes one CSV row; the header row is the union of all flattened
+/// keys, in the order they are first encountered (key order is preserved). Missing keys are
+/// written as empty fields, and `null`/empty values become empty fields.
+///
+/// This is a focused reimplementation of the parts of the unmaintained `json-objects-to-csv`
+/// crate that qsv relied on, so we no longer depend on it (see issue #3523). qsv builds
+/// `serde_json` with the `preserve_order` feature, so object keys are already in insertion order.
+fn json_objects_to_csv<W: Write>(
+    objects: &[Value],
+    mut csv_writer: csv::Writer<W>,
+) -> CliResult<()> {
+    if objects.is_empty() {
+        return Ok(());
+    }
+
+    // First pass: collect the header union in first-seen order, and detect collisions that
+    // would occur when the internal separator is normalized back to `.` (e.g. a genuinely
+    // nested `a.b` key colliding with a literal "a.b" key).
+    let mut headers: Vec<String> = Vec::new();
+    let mut seen_headers: HashSet<String> = HashSet::new();
+    let mut orig_keys: HashSet<String> = HashSet::new();
+    let mut flat = serde_json::Map::new();
+    for obj in objects {
+        flat.clear();
+        flatten_json_value(obj, String::new(), 0, &mut flat)?;
+        for orig_key in flat.keys() {
+            orig_keys.insert(orig_key.clone());
+            let header = transform_flattened_key(orig_key);
+            if seen_headers.insert(header.clone()) {
+                headers.push(header);
+            }
+        }
+    }
+
+    // nothing to write if no headers were produced (e.g. all-empty objects)
+    if headers.is_empty() {
+        return Ok(());
+    }
+    // if two distinct flattened keys normalize to the same header, it's a collision
+    if headers.len() != orig_keys.len() {
+        return fail_clierror!(
+            "Flattening Key Collision error: two different JSON keys collide after flattening"
+        );
+    }
+
+    csv_writer.write_record(&headers)?;
+
+    // Second pass: write one CSV row per object.
+    let mut record: Vec<String> = Vec::with_capacity(headers.len());
+    let mut transformed: serde_json::Map<String, Value> = serde_json::Map::new();
+    for obj in objects {
+        flat.clear();
+        flatten_json_value(obj, String::new(), 0, &mut flat)?;
+
+        transformed.clear();
+        for (orig_key, val) in &flat {
+            transformed.insert(transform_flattened_key(orig_key), val.clone());
+        }
+
+        record.clear();
+        for header in &headers {
+            match transformed.get(header) {
+                Some(Value::String(s)) => record.push(s.clone()),
+                Some(v @ (Value::Bool(_) | Value::Number(_))) => record.push(v.to_string()),
+                // Null, plus any empty Array/Object that slipped through, become empty fields.
+                _ => record.push(String::new()),
+            }
+        }
+        csv_writer.write_record(&record)?;
+    }
+
+    csv_writer.flush()?;
+    Ok(())
 }
 
 /// Convert a jaq `Val` to a `serde_json::Value` without going through Display.

--- a/tests/test_100/lesson_3.rs
+++ b/tests/test_100/lesson_3.rs
@@ -6,7 +6,6 @@ use crate::workdir::Workdir;
 // Task 1
 #[test]
 #[cfg(not(feature = "datapusher_plus"))]
-// #[ignore = "ignore this test for now as it's flaky coz of json_objects_to_csv crate issues"]
 fn flowers_json_to_csv() {
     let wrk = Workdir::new("flowers_json_to_csv");
     let flowers_json_file = wrk.load_test_file("flowers.json");
@@ -32,7 +31,6 @@ fn flowers_json_to_csv() {
 // Task 2
 #[test]
 #[cfg(not(feature = "datapusher_plus"))]
-// #[ignore = "ignore this test for now as it's flaky coz of json_objects_to_csv crate issues"]
 fn flowers_nested_json_to_csv() {
     let wrk = Workdir::new("flowers_nested_json_to_csv");
     let flowers_nested_json_file = wrk.load_test_file("flowers_nested.json");
@@ -57,7 +55,6 @@ fn flowers_nested_json_to_csv() {
 // Task 3
 #[test]
 #[cfg(not(feature = "datapusher_plus"))]
-// #[ignore = "ignore this test for now as it's flaky coz of json_objects_to_csv crate issues"]
 fn buses_csv_to_json() {
     let wrk = Workdir::new("buses_csv_to_json");
     let buses_csv_file = wrk.load_test_file("buses.csv");

--- a/tests/test_json.rs
+++ b/tests/test_json.rs
@@ -436,6 +436,24 @@ fn json_flatten_key_collision() {
 
 #[test]
 #[serial]
+// a key that literally contains the (former) internal separator char (U+241D) must be preserved
+// verbatim, not corrupted into a `.`-separated key (regression guard for the structural flattener)
+fn json_literal_separator_char_preserved() {
+    let wrk = Workdir::new("json_literal_separator_char_preserved");
+    // the key is `a␝b` (U+241D between a and b); it must NOT become `a.b`
+    wrk.create_from_string("data.json", "[{\"a\u{241D}b\":1,\"c\":2}]");
+    let mut cmd = wrk.command("json");
+    cmd.arg("data.json");
+
+    wrk.assert_success(&mut cmd);
+
+    let got: Vec<Vec<String>> = wrk.read_stdout(&mut cmd);
+    let expected = vec![svec!["a\u{241D}b", "c"], svec!["1", "2"]];
+    assert_eq!(got, expected);
+}
+
+#[test]
+#[serial]
 fn json_empty_keys_with_jaq() {
     let wrk = Workdir::new("json_empty_keys_with_jaq");
     let json_data = serde_json::json!({

--- a/tests/test_json.rs
+++ b/tests/test_json.rs
@@ -331,6 +331,111 @@ fn json_nested() {
 
 #[test]
 #[serial]
+// nested objects are flattened with a `.` separator into `parent.child` columns
+fn json_nested_object() {
+    let wrk = Workdir::new("json_nested_object");
+    wrk.create_from_string(
+        "data.json",
+        r#"[{"name":"apple","details":{"color":"red","qty":5}},{"name":"banana","details":{"color":"yellow","qty":3}}]"#,
+    );
+    let mut cmd = wrk.command("json");
+    cmd.arg("data.json");
+
+    wrk.assert_success(&mut cmd);
+
+    let got: Vec<Vec<String>> = wrk.read_stdout(&mut cmd);
+    let expected = vec![
+        svec!["name", "details.color", "details.qty"],
+        svec!["apple", "red", "5"],
+        svec!["banana", "yellow", "3"],
+    ];
+    assert_eq!(got, expected);
+}
+
+#[test]
+#[serial]
+// nested arrays are flattened using the element index as the key segment (`tags.0`, `tags.1`)
+fn json_nested_array() {
+    let wrk = Workdir::new("json_nested_array");
+    wrk.create_from_string(
+        "data.json",
+        r#"[{"name":"apple","tags":["fruit","red"]},{"name":"kiwi","tags":["fruit","green"]}]"#,
+    );
+    let mut cmd = wrk.command("json");
+    cmd.arg("data.json");
+
+    wrk.assert_success(&mut cmd);
+
+    let got: Vec<Vec<String>> = wrk.read_stdout(&mut cmd);
+    let expected = vec![
+        svec!["name", "tags.0", "tags.1"],
+        svec!["apple", "fruit", "red"],
+        svec!["kiwi", "fruit", "green"],
+    ];
+    assert_eq!(got, expected);
+}
+
+#[test]
+#[serial]
+// keys missing from some objects become empty fields, and keys not in the first object are
+// appended as additional columns in the order they are first encountered
+fn json_heterogeneous_missing_keys() {
+    let wrk = Workdir::new("json_heterogeneous_missing_keys");
+    wrk.create_from_string(
+        "data.json",
+        r#"[{"fruit":"apple","cost":1.75,"price":2.50},{"fruit":"mangosteen","price":5.00},{"fruit":"starapple","rating":9,"price":4.50}]"#,
+    );
+    let mut cmd = wrk.command("json");
+    cmd.arg("data.json");
+
+    wrk.assert_success(&mut cmd);
+
+    let got: Vec<Vec<String>> = wrk.read_stdout(&mut cmd);
+    let expected = vec![
+        svec!["fruit", "cost", "price", "rating"],
+        svec!["apple", "1.75", "2.5", ""],
+        svec!["mangosteen", "", "5.0", ""],
+        svec!["starapple", "", "4.5", "9"],
+    ];
+    assert_eq!(got, expected);
+}
+
+#[test]
+#[serial]
+// JSON null values are written as empty CSV fields
+fn json_null_value_empty_field() {
+    let wrk = Workdir::new("json_null_value_empty_field");
+    wrk.create_from_string("data.json", r#"[{"a":1,"b":null},{"a":2,"b":3}]"#);
+    let mut cmd = wrk.command("json");
+    cmd.arg("data.json");
+
+    wrk.assert_success(&mut cmd);
+
+    let got: Vec<Vec<String>> = wrk.read_stdout(&mut cmd);
+    let expected = vec![svec!["a", "b"], svec!["1", ""], svec!["2", "3"]];
+    assert_eq!(got, expected);
+}
+
+#[test]
+#[serial]
+// a genuinely-nested key (a.b) colliding with a literal "a.b" key is reported as an error
+fn json_flatten_key_collision() {
+    let wrk = Workdir::new("json_flatten_key_collision");
+    wrk.create_from_string("data.json", r#"[{"a":{"b":1}},{"a.b":2}]"#);
+    let mut cmd = wrk.command("json");
+    cmd.arg("data.json");
+
+    wrk.assert_err(&mut cmd);
+
+    let got = wrk.output_stderr(&mut cmd);
+    assert!(
+        got.contains("Flattening Key Collision error"),
+        "unexpected stderr: {got}"
+    );
+}
+
+#[test]
+#[serial]
 fn json_empty_keys_with_jaq() {
     let wrk = Workdir::new("json_empty_keys_with_jaq");
     let json_data = serde_json::json!({

